### ISSUE TITLE
Make expo-yarn-workspaces public again?

### ISSUE THREAD

**Issue by kevinwolfcr** [expo/expo#2662]
So far the only reliable workaround to use expo with monorepos i've found is the package `expo-yarn-workspaces`. It saved me a lot of time!

## But...
Due to the update to metro, removing `metro/src/blacklist`, it broke the compatibility so after upgrading an app to expo `31.0.0` stopped working.

A (dirty) workaround i did was to copy paste `expo-yarn-workspaces@1.1.0/package.json` and `expo-yarn-workspaces@1.1.0/index.js` and created a private package within my apps monorepo, and it worked, so why keep it private?

**Comment by ide** [comment]
The expo-yarn-workspaces package is solely intended for this repo. You're welcome to use it, but we don't want to set expectations that it's supported outside of this repo since we can't anticipate everyone's use case and the health of this repo is our focus.

**Comment by ivkrpv** [comment]
So, what's the best way to use expo 31.0.0 with yarn workspaces now? Spent whole day and still no results.)

**Comment by ide** [comment]
You could try borrowing the approach used in this repo. There isn't any "official" solution - ymmv.

**Comment by nhuesmann** [comment]
> So, what's the best way to use expo 31.0.0 with yarn workspaces now? Spent whole day and still no results.)

I'm in the same boat—can't get it working for the life of me.

**Comment by nhuesmann** [comment]
@kevinwolfcr Could you go a little more in depth into your workaround? I'm fine with monkey patching for right now, I just really want to get this to work. My error is `Cannot find module 'metro/src/blacklist'`.

**Comment by nhuesmann** [comment]
@kevinwolfcr Never mind, I figured it out. @ivkrpv in case you're still wondering, I was able to follow @kevinwolfcr 's steps. Here's what I did:

1. Create a new folder/package in your yarn workspace. I called mine `expo-yarn-workspaces`
2. Copy the contents of the repo I'm linking into your new folder https://github.com/expo/expo/tree/master/packages/expo-yarn-workspaces
3. In the repo linked in step 2, follow the steps in the README—only, in your devDependencies in package.json make sure the version for `expo-yarn-workspaces` is 1.1.0. Yarn workspaces will use the local package you copied, which will give you version 1.1.0 that contains the working code.
4. Prob want to nuke your node_modules for good measure and run a `yarn install`. Let me know if this works!

**Comment by aleksandervalle** [comment]
@nhuesmann I followed your advice above, but kept getting 

```
child_process.js:637
    throw err;
    ^

Error: spawnSync C:\dev\simpleprm\expo-yarn-workspaces\bin\symlink-necessary-packages.js UNKNOWN
    at Object.spawnSync (internal/child_process.js:990:20)
    at spawnSync (child_process.js:601:24)
    at Object.execFileSync (child_process.js:629:13)
    at Object.<anonymous> (C:\dev\simpleprm\expo-yarn-workspaces\bin\expo-yarn-workspaces.js:18:19)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
error Command failed with exit code 1.
```

It may(?) seem like the invocation of `child_process.execFileSync` in `expo-yarn-workspaces.js` is incorrect. After changing 

```
...
  case 'postinstall':
    child_process.execFileSync(path.join(__dirname, 'symlink-necessary-packages.js'), {
      stdio: 'inherit',
    });
    child_process.execFileSync(path.join(__dirname, 'make-entry-module.js'), {
      stdio: 'inherit',
    });
    break;
...
```
to
```
...
  case 'postinstall':
    child_process.execSync('node', [path.join(__dirname, 'symlink-necessary-packages.js')]);
    child_process.execSync('node', [path.join(__dirname, 'make-entry-module.js')]);
    break;
...
```

everything is fine.

Modified is not functionally equivalent to original, but it should do the trick.

**Comment by tiagob** [comment]
[nohoist](https://yarnpkg.com/blog/2018/02/15/nohoist/) worked for me

In the child `package.json` instead of the root:
```json
"workspaces": {
  "nohoist": ["**"]
}
```

**Comment by Slapbox** [comment]
@tiagob can you clarify which `package.json` you put `nohoist` in? `expo-yarn-workspaces` or in the package that houses your Expo app?

**Comment by cinoss** [comment]
@tiagob yeah, that works for me as well.
@Slapbox it's should be at the `package.json` file at root directory.
Doing this also means no `node_modules` packages sharing, and file size will be as big as when you do multiple repos.

**Comment by gutosanches** [comment]
@tiagob 's solution worked great for me too. @Slapbox in the package that houses your Expo app.

**Comment by kevinwolfcr** [comment]
I don't think that's the best solution since it will put all dependencies on `my-monorepo/packages/my-app/node_modules` rather than `my-monorepo/node_modules`. I have created a private package which solves this on some of my projects, I can publish it tomorrow with usage instructions.

**Comment by gutosanches** [comment]
@kevinwolfcr good point, that'd be great! look forward to it

**Comment by kevinwolfcr** [comment]
I have an open source library which supports expo apps on yarn workspaces: https://github.com/kevinwolfcr/formal

**Comment by brentvatne** [comment]
nice one @kevinwolfcr

**Comment by gutosanches** [comment]
@kevinwolfcr copying what you did with formal did it for me :) (with a cool bonus of enabling react hooks on expo too, nice trick). thank you!

**Comment by kevinwolfcr** [comment]
Glad it helped! :).

**Comment by Slapbox** [comment]
@kevinwolfcr thanks for sharing your great work! 

I thought I'd copied all your work correctly, but I'm seeing `React-Native` installed in my `MobileApp` package AND in the root `node_modules` folder. Do you have any ideas what I might have done wrong here?

Yarn and Node are up to date. I'm using Windows and I have trouble running your script, but that's simple enough to run manually. I can't even get to that point though since lerna is setting things up in such a way that I get duplicate packages, and so haste map collisions.

Any thoughts you have would be awesome, but if you're too busy thanks a ton for what you've already done!

**Comment by ide** [comment]
@kevinwolfcr - Thanks for helping people out here.

It seems like this package would be useful to enough people so I removed the "private" flag on the package and published it to npm (v1.1.0). The expectation of **no commitment to supporting the package outside this repo** is the same for now, but there isn't too much that is repo-specific to the package.

If the package is useful to you, it's on npm. If not but the code is useful, it's released under the MIT License and you can take it with respect to the license.

**Comment by stephenhandley** [comment]
In case anyone also has trouble getting `expo-yarn-workspaces` working after following this package's readme, specifically getting the following when trying to import a module with JSX from a shared package:
```
Module parse failed: Unexpected token (4:9)
You may need an appropriate loader to handle this file type.
```

Saw that the internal expo packages using this lib have a typescript build process and realized there needs to be a babel build step for the shared packages. Here's a working example:
https://github.com/stephenhandley/expoyarnlerna
